libpod/boltdb_state_internal.go

```diff
@@ -0,0 +1,194 @@
+package libpod
+
+import (
+	"encoding/json"
+	"path/filepath"
+
+	"github.com/boltdb/bolt"
+	"github.com/containers/storage"
+	"github.com/pkg/errors"
+)
+
+const (
+	idRegistryName    = "id-registry"
+	nameRegistryName  = "name-registry"
+	ctrConfigName     = "container-config"
+	ctrStateName      = "container-state"
+	netNSName         = "net-ns"
+	runtimeConfigName = "runtime-config"
+)
+
+var (
+	idRegistryBkt    = []byte(idRegistryName)
+	nameRegistryBkt  = []byte(nameRegistryName)
+	ctrConfigBkt     = []byte(ctrConfigName)
+	ctrStateBkt      = []byte(ctrStateName)
+	netNSBkt         = []byte(netNSName)
+	runtimeConfigBkt = []byte(runtimeConfigName)
+)
+
+// Check if the configuration of the database is compatible with the
+// configuration of the runtime opening it
+// If there is no runtime configuration loaded, load our own
+func checkRuntimeConfig(db *bolt.DB, runtime *Runtime) error {
+	var (
+		staticDir       = []byte("static-dir")
+		tmpDir          = []byte("tmp-dir")
+		runRoot         = []byte("run-root")
+		graphRoot       = []byte("graph-root")
+		graphDriverName = []byte("graph-driver-name")
+	)
+
+	err := db.Update(func(tx *bolt.Tx) error {
+		configBkt, err := getRuntimeConfigBucket(tx)
+		if err != nil {
+			return err
+		}
+
+		if err := validateDBAgainstConfig(configBkt, "static dir",
+			runtime.config.StaticDir, staticDir); err != nil {
+			return err
+		}
+
+		if err := validateDBAgainstConfig(configBkt, "tmp dir",
+			runtime.config.TmpDir, tmpDir); err != nil {
+			return err
+		}
+
+		if err := validateDBAgainstConfig(configBkt, "run root",
+			runtime.config.StorageConfig.RunRoot, runRoot); err != nil {
+			return err
+		}
+
+		if err := validateDBAgainstConfig(configBkt, "graph root",
+			runtime.config.StorageConfig.GraphRoot, graphRoot); err != nil {
+			return err
+		}
+
+		return validateDBAgainstConfig(configBkt, "graph driver name",
+			runtime.config.StorageConfig.GraphDriverName,
+			graphDriverName)
+	})
+
+	return err
+}
+
+// Validate a configuration entry in the DB against current runtime config
+// If the given configuration key does not exist it will be created
+func validateDBAgainstConfig(bucket *bolt.Bucket, fieldName, runtimeValue string, keyName []byte) error {
+	keyBytes := bucket.Get(keyName)
+	if keyBytes == nil {
+		if err := bucket.Put(keyName, []byte(runtimeValue)); err != nil {
+			return errors.Wrapf(err, "error updating %s in DB runtime config", fieldName)
+		}
+	} else {
+		if runtimeValue != string(keyBytes) {
+			return errors.Wrapf(ErrDBBadConfig, "database %s %s does not match our %s %s",
+				fieldName, string(keyBytes), fieldName, runtimeValue)
+		}
+	}
+
+	return nil
+}
+
+func (s *BoltState) getDBCon() (*bolt.DB, error) {
```

Comment:
The code keeps re-opening/re-closing the DB, so presumably multiples processes can access the DB file concurrently.

In that case, couldn’t the concurrency benefit from the read-only users using `bolt.Open(…, &bolt.Options{ReadOnly: true})`? Or is the structure of the concurrent readers/writers known and such that there can never be two readers in parallel?

Reply:
I looked into that, and the problem is that Bolt's locking code assumes that opening a read-only DB in write mode is an unrecoverable error. We could work around this by repeatedly attempting to open the database or modifying the Bolt code to ignore incompatible lock types and spin until it can acquire the lock, but that sounds like a future optimization

Reply:
I was thinking there would be a `getDBConRW` and `getDBConRO`, each with only one `bolt.Open` call. Or perhaps I misunderstand the problem.

Reply:
The problem would come when we have two separate podman processes trying to access the database, one trying to exclusive lock for write and one shared lock for read-only. Whichever does not get the lock first will error as the lock is in an incompatible state. We can probably work around this by using some form of external locking (a writer lock that must be held to lock the database for write, maybe?) but that sounds like significant extra work that would be occur in a followon PR

Reply:
> The problem would come when we have two separate podman processes trying to access the database, one trying to exclusive lock for write and one shared lock for read-only. Whichever does not get the lock first will error as the lock is in an incompatible state. 

That’s not what I see with the test programs below: `bolt.flock` uses `syscall.Flock`, and blocks, not fails, on a lock collision (`syscall.EWOULDBLOCK`). Whoever does not get the lock stays in the `Opening ...` state, and after the owner of the lock (or all simultaneous readers) close the db, a new blocked owner is unblocked and gets an open database.

(I do see that the order the blocked processes are unblocked is not FIFO, but that should only be an issue under high contention, and anyway is already an issue for the current RW-always locking code.)

```go
package main

import (
	"fmt"
	"github.com/boltdb/bolt"
	"io/ioutil"
	"log"
	"os"
)

func main() {
	fmt.Println("Opening...")
// One of
	db, err := bolt.Open("db", 0600, nil)
// Or
	db, err := bolt.Open("db", 0600, &bolt.Options{ReadOnly: true})
	fmt.Println("... Open")
	if err != nil {
		log.Fatalf("bolt.Open: %#v", err)
	}
	defer db.Close()
	_, _ = ioutil.ReadAll(os.Stdin) // Wait indefinitely for Ctrl-D
	fmt.Println("Closing")
}
```

Reply:
Hm. I'll look into this. This would definitely be a major optimization if we can get concurrent reads working.